### PR TITLE
Feat: add dynamic generation

### PR DESCRIPTION
**Pull request by saenyakorn**

### CODE REVIEW

File: components/announcements/recruiting.jsx [hunk 1]
```diff
@@ -1,12 +1,39 @@
+import differenceInYears from 'date-fns/differenceInYears'
+
+function ordinal(num) {
+  num = Math.round(num)
+  let numString = num.toString()
+
+  if (Math.floor(num / 10) % 10 === 1) {
```
Comment (new5558) [inline components/announcements/recruiting.jsx]: Can we just use date-fns to format this instead of writing your own function

Comment (bombnp) [reply components/announcements/recruiting.jsx]: if we decide to use our own function, i suggest we move this somewhere else as a util func

File: components/announcements/recruiting.jsx [hunk 2]
```diff
@@ -31,7 +58,8 @@ const Recruiting = () => {
                 </a>
               </div>
               <p className="text-sm text-gray-600 text-center font-normal leading-loose">
-                or search for "<span className="font-bold">Thinc. 9 Gen</span>" in your FB application!
+                or search for "<span className="font-bold">Thinc. {ordinal(currentThincGeneration)} Gen</span>" in your
+                FB application!
```
Comment (bombnp) [inline components/announcements/recruiting.jsx]: we're gonna assume we use the same name format every year?

**Comment by new5558** [comment]
image
Turn out nextjs 8.1.0 static site generator automatically generate the number from `new Date()` at build time, so this PR is not *dynamic as it should be. I think [dynamic import](https://nextjs.org/docs/advanced-features/dynamic-import) should solve this problem.

File: components/announcements/recruiting.jsx (hunk 1 above)
Comment (saenyakorn) [reply components/announcements/recruiting.jsx]: Unfortunately, date-fns don't have `ordinal` or other relate function. Need to add more library.

Comment (new5558) [reply components/announcements/recruiting.jsx]: image
It [has](https://date-fns.org/v2.25.0/docs/format)

File: components/announcements/recruiting.jsx (hunk 2 above)
Comment (saenyakorn) [reply components/announcements/recruiting.jsx]: Initially, we will assume that we use the same name format for FB group. But if nong want to change the format, they can just create a new PR.

**Comment by saenyakorn** [comment]
> dynamic import

dynamic import what?

**Comment by new5558** [comment]
> > dynamic import
> 
> dynamic import what?

```
import dynamic from 'next/dynamic'

const DynamicComponentWithNoSSR = dynamic(
  () => import('../components/hello3'),
  { ssr: false }
)
```

**Comment by bombnp** [comment]
personally i don't mind if it's not dynamic.
it's not hard at all to change gen

File: components/announcements/recruiting.jsx (hunk 1 above)
Comment (saenyakorn) [reply components/announcements/recruiting.jsx]: So, how to get 32nd?

File: pages/index.jsx [hunk 3]
```diff
@@ -13,6 +12,10 @@ const Members = dynamic(() => import('../components/sections/members'), {
   ssr: false,
 })
 
+const Recruiting = dynamic(() => import('../components/announcements/recruiting'), {
```
Comment (new5558) [inline pages/index.jsx]: Preview: https://thinc-www-new17353.vercel.app/
Ok, now the dynamic number is working but react needs to render recruiting section after other sections are loaded on each page load. This creates severe [cumulative layout shift](https://web.dev/cls/)

image
image

File: components/announcements/recruiting.jsx [hunk 4]
```diff
@@ -1,12 +1,20 @@
...
+
   return (
     <section className="w-full bg-pink-100 py-20">
       <div className="container">
         <div className="mx-8">
           <div className="w-full flex justify-center flex-row">
             <div className="w-full flex flex-col items-center md:w-2/3">
               <p className="text-4xl text-center font-headline font-semibold leading-none text-black mb-10">
-                Hello, CU105
+                Hello, CU{currentCUGeneration}
```
Comment (new5558) [inline components/announcements/recruiting.jsx]: Can we create new text component and use dynamic just for this number

File: components/announcements/recruiting.jsx [hunk 5]
```diff
@@ -31,7 +39,8 @@ const Recruiting = () => {
                 </a>
               </div>
               <p className="text-sm text-gray-600 text-center font-normal leading-loose">
-                or search for "<span className="font-bold">Thinc. 9 Gen</span>" in your FB application!
+                or search for "<span className="font-bold">Thinc. {ordinal(currentThincGeneration)} Gen</span>" in your
```
Comment (new5558) [inline components/announcements/recruiting.jsx]: Also dynamic this number

File: pages/index.jsx (hunk 3 above)
Comment (new5558) [reply pages/index.jsx]: Sorry for not very elaborate at first krub. You can look at my suggestions at the below comments

**Comment by new5558** [comment]
image
I think we can still do better if we only put dynamic number in the component that will be dynamic imported and use loading placeholder.

File: components/announcements/recruiting.jsx [hunk 6]
```diff
@@ -1,13 +1,21 @@
+import dynamic from 'next/dynamic'
+
+const CUGeneration = dynamic(() => import('./cuGeneration'), {
+  ssr: false,
+})
+
+const ThincGeneration = dynamic(() => import('./thincGeneration'), {
```
Comment (new5558) [inline components/announcements/recruiting.jsx:7]:
```
const ThincGeneration = dynamic(() => import('./thincGeneration'), {
  ssr: false,
  loading: () => <>loading...</>
})
```

File: components/announcements/recruiting.jsx [hunk 7]
```diff
@@ -1,13 +1,21 @@
+import dynamic from 'next/dynamic'
+
+const CUGeneration = dynamic(() => import('./cuGeneration'), {
```
Comment (new5558) [inline components/announcements/recruiting.jsx:3]:
```
const CUGeneration = dynamic(() => import('./cuGeneration'), {
  ssr: false,
  loading: () => <>loading...</>
})
```

File: components/announcements/cuGeneration.jsx [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { differenceInYears } from 'date-fns'
+
+const CUGeneration = () => {
+  // Assume: increase generation every 31st July of every year
+  // Note: Javascript count January as 0
+  const currentCUGeneration = differenceInYears(new Date(), new Date(1916, 6, 31))
+  return (
```
Comment (new5558) [inline components/announcements/cuGeneration.jsx]: `return <>{currentCUGeneration}</>`

**Comment by saenyakorn** [comment]
Why don't we just use ServerSideRendering. This problem must not happen.

File: components/announcements/thincGeneration.jsx [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import { differenceInYears } from 'date-fns'
+import ordinal from 'ordinal'
+
+const ThincGeneration = () => {
+  // Assume: increase generation every 31st July of every year
+  // Note: Javascript count January as 0
+  const currentCUGeneration = differenceInYears(new Date(), new Date(1916, 6, 31))
+  const currentThincGeneration = currentCUGeneration - 96
+  return (
```
Comment (new5558) [inline components/announcements/thincGeneration.jsx]: `return <>{currentThincGeneration}</>`

File: components/announcements/thincGeneration.jsx [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import { differenceInYears } from 'date-fns'
+import ordinal from 'ordinal'
+
+const ThincGeneration = () => {
+  // Assume: increase generation every 31st July of every year
+  // Note: Javascript count January as 0
+  const currentCUGeneration = differenceInYears(new Date(), new Date(1916, 6, 31))
```
Comment (new5558) [inline components/announcements/thincGeneration.jsx:7]: remove this

File: components/announcements/recruiting.jsx [hunk 11]
```diff
@@ -1,13 +1,21 @@
...
   return (
     <section className="w-full bg-pink-100 py-20">
       <div className="container">
         <div className="mx-8">
           <div className="w-full flex justify-center flex-row">
             <div className="w-full flex flex-col items-center md:w-2/3">
-              <p className="text-4xl text-center font-headline font-semibold leading-none text-black mb-10">
-                Hello, CU105
-              </p>
+              <CUGeneration />
```
Comment (new5558) [inline components/announcements/recruiting.jsx]:
```
 <p className="text-4xl text-center font-headline font-semibold leading-none text-black mb-10">
    Hello, CU<CUGeneration />
</p>
```

File: components/announcements/recruiting.jsx [hunk 12]
```diff
@@ -30,16 +38,13 @@ const Recruiting = () => {
                   VISIT <span className="ml-1 font-bold"> FACEBOOK PAGE</span>
                 </a>
               </div>
-              <p className="text-sm text-gray-600 text-center font-normal leading-loose">
-                or search for "<span className="font-bold">Thinc. 9 Gen</span>" in your FB application!
-              </p>
+              <ThincGeneration />
```
Comment (new5558) [inline components/announcements/recruiting.jsx]:
```
<p className="text-sm text-gray-600 text-center font-normal leading-loose">
    or search for "<span className="font-bold">Thinc. <ThincGeneration /> Gen</span>" in your FB application!
</p>
```

**Comment by new5558** [comment]
> Why don't we just use ServerSideRendering. This problem must not happen.

~~To be precise, it is static site generator problem not SSR problem. But if you want to migrate to react/next11 would be appreciated.
I am not sure whether next11 have this problem or not?~~

we have free server?

File: components/announcements/recruiting.jsx (hunk 6 above)
Comment (saenyakorn) [reply components/announcements/recruiting.jsx:7]: use `loading...` before getting the generation is not a good UX/UI. Just leave it empty

**Review by new5558 — APPROVED** [review]
LGTM, I don't think we can do more from this. any comments?
image

**Pull request merged**